### PR TITLE
strings: fix type hints and uncovered bugs

### PR DESCRIPTION
**Pull request by williballenthin**
When adding type hints (at the suggestion of my editor), I found that some of the strings.py helpers probably didn't ever work on python3. So I fixed those up and added some tests.

### TODO

- [x] changelog

- [x] tests

- [x] No documentation update needed

### CODE REVIEW

**Comment by mr-tz** [comment]
Also do this for FLOSS?

**Review by mike-hunhoff — APPROVED** [review]
Looks good, thanks for the tests! LGTM 🚀

File: capa/features/extractors/strings.py [hunk 1]
```diff
@@ -11,38 +11,71 @@
...
+    s: str
+    offset: int
+
+
+def buf_filled_with(buf: bytes, character: int) -> bool:
+    """Check if the given buffer is filled with the given character, repeatedly.
+
+    Args:
+        buf: The bytes buffer to check
+        character: The byte value (0-255) to check for
```
Comment (fariss) [inline capa/features/extractors/strings.py:38]: It might be useful to add a check to `buf_filled_with` to assert `character` in the (0-255) range
```python
    if not (0 <= character <= 255):
        raise ValueError(f"Character value {character} outside valid byte range (0-255)")
```

File: capa/features/extractors/strings.py [hunk 2]
```diff
@@ -11,38 +11,71 @@
...
     Extract ASCII strings from the given binary data.
 
-    :param buf: A bytestring.
-    :type buf: str
-    :param n: The minimum length of strings to extract.
-    :type n: int
-    :rtype: Sequence[String]
+    Params:
+      buf: the bytes from which to extract strings
+      n: minimum string length
```
Comment (fariss) [inline capa/features/extractors/strings.py:78]: Similary, check that lengh `n` is valid:
```py
def extract_ascii_strings(buf: bytes, n: int = 4) -> Iterator[String]:
  if n < 1:
    raise ValueError("minimum length must be positive")
```

File: tests/test_strings.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    # mixed bytes
+    assert buf_filled_with(b"\x00\x01" * 8, 0x00) is False
+    assert buf_filled_with(b"ABCD" * 8, ord("A")) is False
+
+    # edge cases
+    assert buf_filled_with(b"", 0x00) is False  # Empty buffer
+    assert buf_filled_with(b"\x00", 0x00) is True  # Single byte
+
+
+def test_extract_ascii_strings():
```
Comment (fariss) [inline tests/test_strings.py:32]: Some edge cases to consider for `test_extract_ascii_strings`:
```py
    # test empty buffer
    assert list(extract_ascii_strings(b"")) == []
    
    # Test buffer with only non-ASCII
    assert list(extract_ascii_strings(b"\xFF\xFF\xFF")) == []
```

**Review by fariss — APPROVED** [review]
Good improvements, thanks. I have left some comments for your review.

**Comment by williballenthin** [comment]
these are all great ideas @fariss thank you! i'll add those in the morning and merge. thanks all

**Review by yelhamer — APPROVED** [review]
LGTM, Thanks!

File: capa/features/extractors/strings.py [hunk 4]
```diff
@@ -11,38 +11,71 @@
...
 import contextlib
-from collections import namedtuple
+from dataclasses import dataclass
+from collections.abc import Iterator
 
 ASCII_BYTE = r" !\"#\$%&\'\(\)\*\+,-\./0123456789:;<=>\?@ABCDEFGHIJKLMNOPQRSTUVWXYZ\[\]\^_`abcdefghijklmnopqrstuvwxyz\{\|\}\\\~\t".encode(
     "ascii"
 )
 ASCII_RE_4 = re.compile(b"([%s]{%d,})" % (ASCII_BYTE, 4))
 UNICODE_RE_4 = re.compile(b"((?:[%s]\x00){%d,})" % (ASCII_BYTE, 4))
```
Comment (fariss) [inline capa/features/extractors/strings.py:21]: imo `4` here is not very descriptive, I'd suggest declaring it a constant:
```py
MIN_STRING_LENGTH = 4 

ASCII_RE_4 = re.compile(b"([%s]{%d,})" % (ASCII_BYTE, MIN_STRING_LENGTH))
UNICODE_RE_4 = re.compile(b"((?:[%s]\x00){%d,})" % (ASCII_BYTE, MIN_STRING_LENGTH))
```

**Comment by williballenthin** [comment]
got the whole crew on this thread!

👋
